asf_public_discourse_home_decarbonisation/analysis/mse/eda_mse_category_data.py

```diff
@@ -0,0 +1,311 @@
+    )
+
+    logger.info(
+        f"Data collected ranges from {mse_data['datetime'].min()} to {mse_data['datetime'].max()}"
+    )
+
+
+def plot_number_posts_per_year(mse_data: pd.DataFrame, category: str):
+    """
+    Plots and saves a bar plot with number of posts per year.
+
+    Args:
+        mse_data (pd.DataFrame): Money Saving Expert data
+        category (str): MSE category/sub-forum
+    """
+    posts_per_year = mse_data.groupby("year")[["id"]].nunique()
+    posts_per_year.columns = ["number_posts"]
+    posts_per_year.reset_index(inplace=True)
+
+    chart = (
+        alt.Chart(posts_per_year)
+        .mark_bar()
+        .encode(
+            y=alt.Y("year:O", title="Year"),
+            x=alt.X("number_posts:Q", title="Number of posts"),
+            color=alt.value(pc.NESTA_COLOURS[1]),
+        )
+        .properties(width=400, height=300)
+        .interactive()
+    )
+    chart = pc.configure_plots(chart, "Number of posts per year", "")
+    path_to_plot = f"{MSE_FIGURES_PATH}/category_{category}_posts_per_year.html"
+    chart.save(path_to_plot)
+
+    logger.info(
+        f"Plot with number of posts per year for category '{category}' saved successfully in {path_to_plot}"
+    )
+
+
+def check_username_id_correspondence(mse_data: pd.DataFrame):
+    """
+    Checks if there is a 1-1 correspondence between `username` and `user_id`.
+
+    Args:
+        mse_data (pd.DataFrame): MSE category/sub-forum
+    """
+    check_id_username = mse_data.groupby("username")[["user_id"]].nunique()
+
+    usernames_corresponding_to_multiple_ids = check_id_username[
+        check_id_username["user_id"] > 1
+    ]
+
+    if len(usernames_corresponding_to_multiple_ids) > 0:
+        logger.warning(
+            f"There are {len(usernames_corresponding_to_multiple_ids)} username(s) corresponding to multiple user_ids."
+        )
+        if set(usernames_corresponding_to_multiple_ids.index) == {"[Deleted User]"}:
+            logger.info(
+                "All usernames corresponding to multiple user_ids are Deleted Users."
+            )
+        else:
+            logger.warning(
+                "There are usernames corresponding to multiple user_ids that are not Deleted Users."
+            )
+
+    check_id_username2 = mse_data.groupby("user_id")[["username"]].nunique()
+
+    ids_corresponding_to_multiple_usernames = check_id_username2[
+        check_id_username2["username"] > 1
+    ]
+    if len(ids_corresponding_to_multiple_usernames) > 0:
+        logger.warning(
+            f"There are {len(ids_corresponding_to_multiple_usernames)} user IDs corresponding to multiple usernames."
+        )
+
+
+def plot_number_interactions_per_user(mse_data: pd.DataFrame, category: str):
+    """
+    Plots and saves a histogram with number of interactions per user.
+    The computation is based on all MSE posts, not necessarily the ones we collected data for.
+
+    Args:
+        mse_data (pd.DataFrame): MSE category/sub-forum
```

Comment:
```suggestion
        mse_data (pd.DataFrame): Money Saving Expert data
```

Reply:
fixed this one and a few others, thank you